site/blocks/StickyNav.tsx

```diff
@@ -0,0 +1,179 @@
+import React, { createRef } from "react"
+import ReactDOM from "react-dom"
+import classnames from "classnames"
+import { debounce, get } from "../../clientUtils/Util.js"
+
+function getTotalOffset(element: HTMLElement): {
+    x: number
+    y: number
+} {
+    let y = 0
+    let x = 0
+    while (true) {
+        x += element.offsetLeft
+        y += element.offsetTop
+        if (!element.offsetParent) {
+            break
+        }
+        element = element.offsetParent as HTMLElement
+    }
+    return { x, y }
+}
+
+const STICKY_NAV_HEIGHT = 54
+const HEADING_SCROLL_MARGIN = 16
+
+interface StickyNavProps {
+    links: { text: string; target: string }[]
+}
+class StickyNav extends React.Component<
+    StickyNavProps,
+    {
+        headingPositions: { id: string; position: number }[]
+        currentHeadingIndex?: number
+    }
+> {
+    ulRef = createRef<HTMLUListElement>()
+
+    state = {
+        currentHeadingIndex: undefined,
+        headingPositions: [] as { id: string; position: number }[],
+    }
+
+    get links() {
+        return this.props.links
+    }
+
+    handleResize = debounce(() => {
+        this.setHeadingPositions()
+    }, 100)
+
+    handleScroll = debounce(() => {
+        this.setCurrentHeading()
+    }, 100)
```

Comment:
Using `debounce` here means that, if I never stop scrolling, then it will never update the current heading. You can use lodash's `throttle` instead.
For a slightly more in-depth explanation and demo, see here for the difference: https://stackoverflow.com/a/25991510/10670163

Reply:
Yeah it can often be difficult to choose between the two strategies. 🙂 

I went with debounce because using throttle means that if the last call is made just before you scroll over the threshold to a new heading, the wrong value will be set.

I think it's better to know that the heading will always be correct than to have it animate as you scroll but potentially land in the wrong state. I'll double check with Matt that he's okay with this.

Reply:
We switched to throttle 👍